### ISSUE TITLE
Clarify offline plugin instructions from plugin command

### ISSUE THREAD

**Issue by ppf2** [elastic/logstash#4651]
To reproduce, create 2 separate LS 2.1 directories, one for the logstash-2.1.0_source (with internet access to prepare the offline package), the other logstash-2.1.0_target (to unpack the offline package).

First prepare the offline package in logstash-2.1.0_source:

```
./plugin update logstash-input-kafka 
./plugin update logstash-output-elaticsearch 
./plugin pack 
```

In the logstash-2.1.0_target location:

```
./plugin list --verbose |grep kafka
logstash-input-kafka (2.0.2)
logstash-output-kafka (2.0.1)
```

```
./plugin unpack ../../logstash-2.1.0_source/plugins_package.tar.gz 
Unpacking ../../logstash-2.1.0_source/plugins_package.tar.gz
Unpacked at /Users/test_user/Downloads/testing/logstash-2.1.0_target/vendor/cache
The unpacked plugins can now be installed in local-only mode using bin/plugin install --local [plugin name]
```

See how it tells you to use the `install --local plugin_name` option for the plugin command.

If you try this (now with internet access disabled), you will see that the plugin install using the local cache will fail:

```
./plugin install --local logstash-input-kafka
Installing logstash-input-kafka
Error Bundler::HTTPError, retrying 1/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 2/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 3/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 4/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 5/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 6/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 7/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 8/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 9/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 10/10
Could not fetch specs from https://rubygems.org/
Too many retries, aborting, caused by Bundler::HTTPError
ERROR: Installation Aborted, message: Could not fetch specs from https://rubygems.org/
```

Same thing if you use the update command:

```
./plugin update --local logstash-input-kafka
Updating logstash-input-kafka
Error Bundler::HTTPError, retrying 1/10
Could not fetch specs from https://rubygems.org/
Error Bundler::HTTPError, retrying 2/10
Could not fetch specs from https://rubygems.org/
```

The following seems to work as a workaround by using the `update` action _and_ just `--local` to update all packages:

```
./plugin update --local
Updating logstash-codec-collectd, logstash-codec-dots, logstash-codec-edn, logstash-codec-edn_lines, logstash-codec-es_bulk, logstash-codec-fluent, logstash-codec-graphite, logstash-codec-json, logstash-codec-json_lines, logstash-codec-line, logstash-codec-msgpack, logstash-codec-multiline, logstash-codec-netflow, logstash-codec-oldlogstashjson, logstash-codec-plain, logstash-codec-rubydebug, logstash-filter-anonymize, logstash-filter-checksum, logstash-filter-clone, logstash-filter-csv, logstash-filter-date, logstash-filter-dns, logstash-filter-drop, logstash-filter-fingerprint, logstash-filter-geoip, logstash-filter-grok, logstash-filter-json, logstash-filter-kv, logstash-filter-metrics, logstash-filter-multiline, logstash-filter-mutate, logstash-filter-ruby, logstash-filter-sleep, logstash-filter-split, logstash-filter-syslog_pri, logstash-filter-throttle, logstash-filter-urldecode, logstash-filter-useragent, logstash-filter-uuid, logstash-filter-xml, logstash-input-beats, logstash-input-couchdb_changes, logstash-input-elasticsearch, logstash-input-eventlog, logstash-input-exec, logstash-input-file, logstash-input-ganglia, logstash-input-gelf, logstash-input-generator, logstash-input-graphite, logstash-input-heartbeat, logstash-input-http, logstash-input-imap, logstash-input-irc, logstash-input-jdbc, logstash-input-kafka, logstash-input-log4j, logstash-input-lumberjack, logstash-input-pipe, logstash-input-rabbitmq, logstash-input-redis, logstash-input-s3, logstash-input-snmptrap, logstash-input-sqs, logstash-input-stdin, logstash-input-syslog, logstash-input-tcp, logstash-input-twitter, logstash-input-udp, logstash-input-unix, logstash-input-xmpp, logstash-input-zeromq, logstash-output-cloudwatch, logstash-output-csv, logstash-output-elasticsearch, logstash-output-email, logstash-output-exec, logstash-output-file, logstash-output-ganglia, logstash-output-gelf, logstash-output-graphite, logstash-output-hipchat, logstash-output-http, logstash-output-irc, logstash-output-juggernaut, logstash-output-kafka, logstash-output-lumberjack, logstash-output-nagios, logstash-output-nagios_nsca, logstash-output-null, logstash-output-opentsdb, logstash-output-pagerduty, logstash-output-pipe, logstash-output-rabbitmq, logstash-output-redis, logstash-output-s3, logstash-output-sns, logstash-output-sqs, logstash-output-statsd, logstash-output-stdout, logstash-output-tcp, logstash-output-udp, logstash-output-xmpp, logstash-output-zeromq
Updated logstash-output-elasticsearch 2.1.4 to 2.3.0
```

```
./plugin list --verbose |grep kafka
logstash-input-kafka (2.0.4)
logstash-output-kafka (2.0.1)
```

There are 2 improvements here we can make to the plugins command:
- `The unpacked plugins can now be installed in local-only mode using bin/plugin install --local [plugin name]` is a misleading message since it doesn't actually work
- If you look at the output of the workaround `./plugin update --local`, you will see that it only provides an indication that it updated the ES output, but for some reason, no indication that it updated the kafka input (the plugin list command later on confirms it though).

**Comment by purbon** [comment]
@ppf2 I understand we should clarify this, happy to help. Can you share with my how your Gemfile  and Gemfile.lock files looks like?

**Comment by michalterbert** [comment]
I have the same issue.

[Gemfile.jruby-1.9.lock.txt](https://github.com/elastic/logstash/files/123029/Gemfile.jruby-1.9.lock.txt)
[Gemfile.txt](https://github.com/elastic/logstash/files/123030/Gemfile.txt)

**Comment by vmcano** [comment]
I also have the same issue when trying to update offline `logstash-input-beats` plugin to solve the issue described in https://github.com/logstash-plugins/logstash-input-beats/issues/46.

**Comment by purbon** [comment]
I noticed the same kind of errors, trying to narrow down probable cause right now.

**Comment by purbon** [comment]
Hi,
 I think I got to the bottom of this problem, the situation is that when doing the updates and then running the `pack` command, the internal cache build has then some missing gems required when doing the update/install process in the offline machine. Path to fix this issue will be:

* Make sure install/update command does not remove need gems.
* Deal with such a situation when installing plugins and report a more clean error.

**Comment by purbon** [comment]
After giving it some thoughts I create #4779 with a bit of improvement here, basically this make the errors more clear. In the PR description I also describe how I think people should drive, for now, such a situation, copying here for reference:


>This situation is raised when the user does, while working with an old version (not necessary too old, but that uses old dependencies), see #4651 for details:
* Run bin/plugin update [plugin name]. < could be with more than one plugin, or just one>
* Run bin/plugin pack

>If for example plugin name depends on A version 1.0, but the new version actually can use a new version, lets say 2.0, the package will contain only 2.0 as is the current state in the LS used to create the bundle. This will cause problems when using this package in a raw LS, there bundler might think it needs 1.0 version. This might also happen if 2 or more plugins are updated, due to internal cache checkups done in Bundler, see https://github.com/bundler/bundler/blob/master/lib/bundler/cli/update.rb#L53 for more details.

> To address this situation I recommend, for now and without disregarding feature improvements, to:
* Create an initial pack before doing the update. Name this file something like initial_plugin_package.tar.gz, so there are no override in the feature.
* Update the desired plugins.
* Create a new package, that will include the new plugins.
* Merge the initial and the new package into a single package.

>The merged package will include old and new dependencies, making sure all required dependencies are available to the offline environment. Complementary improvements in the document should be made, so this is known. This is useful if you only want to update 1 plugin, but got more updated in your package.

>There is also the situation when you know and control which plugins are new in your package for offline installation, then I recommend to use bin/plugin update --local as this will make sure bundler does the right internal dependency resolution.

@ppf2 @suyograo @ph feedback welcome, I know the process might be a bit hard to follow, don't hesitate to ask for more details :+1:

**Comment by colinsurprenant** [comment]
I added my thoughts on this issue in #4779.

**Comment by colinsurprenant** [comment]
I moved to discussion to #4787

**Comment by ppf2** [comment]
Can I get a clarification on this one:

1.  Is the issue addressed in 2.3.1+?
2.  If so, will there be an update to the documentation on the extra steps required to perform an offline update of a plugin (i.e. https://github.com/elastic/logstash/issues/4651#issuecomment-194223087)?

Thanks,
Pius

**Comment by acchen97** [comment]
Closing, shiny new workflow introduced in 5.2. https://www.elastic.co/guide/en/logstash/current/offline-plugins.html
